cmd/harmony/dumpdb.go

```diff
@@ -0,0 +1,419 @@
+	}
+}
+
+// OnAccount implements DumpCollector interface
+func (db *KakashiDB) OnAccountEnd(addr common.Address, acc state.DumpAccount) {
+	accountCount++
+	accountState = ON_ACCOUNT_END
+	if totalSize-flushSize > int(db.batchLimit) {
+		db.flush()
+	}
+}
+
+func (db *KakashiDB) getHashByNumber(number uint64) common.Hash {
+	hash := rawdb.ReadCanonicalHash(db, number)
+	return hash
+}
+func (db *KakashiDB) GetHeaderByNumber(number uint64) *block.Header {
+	hash := db.getHashByNumber(number)
+	if hash == (common.Hash{}) {
+		return nil
+	}
+	return db.GetHeader(hash, number)
+}
+
+func (db *KakashiDB) GetHeader(hash common.Hash, number uint64) *block.Header {
+	header := rawdb.ReadHeader(db, hash, number)
+	return header
+}
+
+func (db *KakashiDB) GetHeaderByHash(hash common.Hash) *block.Header {
+	number := rawdb.ReadHeaderNumber(db, hash)
+	return rawdb.ReadHeader(db, hash, *number)
+}
+
+// GetBlock retrieves a block from the database by hash and number
+func (db *KakashiDB) GetBlock(hash common.Hash, number uint64) *types.Block {
+	block := rawdb.ReadBlock(db, hash, number)
+	return block
+}
+
+// GetBlockNumber retrieves the block number belonging to the given hash
+// from the database
+func (db *KakashiDB) GetBlockNumber(hash common.Hash) *uint64 {
+	return rawdb.ReadHeaderNumber(db, hash)
+}
+
+// GetBlockByHash retrieves a block from the database by hash
+func (db *KakashiDB) GetBlockByHash(hash common.Hash) *types.Block {
+	number := db.GetBlockNumber(hash)
+	return db.GetBlock(hash, *number)
+}
+
+// GetBlockByNumber retrieves a block from the database by number
+func (db *KakashiDB) GetBlockByNumber(number uint64) *types.Block {
+	hash := rawdb.ReadCanonicalHash(db, number)
+	return db.GetBlock(hash, number)
+}
+
+func (db *KakashiDB) indexerDataDump(block *types.Block) {
+	fmt.Println("indexerDataDump:")
+	bloomIndexer := hmy.NewBloomIndexer(db, params.BloomBitsBlocks, params.BloomConfirms)
+	bloomIndexer.Close() // just stop event loop
+	section, blkno, blkhash := bloomIndexer.Sections()
+	bloomIndexer.AddCheckpoint(section-1, blkhash)
+	for i := blkno; i <= block.NumberU64(); i++ {
+		db.GetHeaderByNumber(i)
+	}
+	db.flush()
+}
+
+func (db *KakashiDB) offchainDataDump(block *types.Block) {
+	fmt.Println("offchainDataDump:")
+	rawdb.WriteHeadBlockHash(db.toDBBatch, block.Hash())
+	rawdb.WriteHeadHeaderHash(db.toDBBatch, block.Hash())
+	db.GetHeaderByNumber(0)
+	db.GetBlockByNumber(0)
+	db.GetHeaderByHash(block.Hash())
+	// EVM may access the last 256 block hash
+	for i := 0; i <= BLOCKS_DUMP; i++ {
+		if block.NumberU64() < uint64(i) {
+			break
+		}
+		latestNumber := block.NumberU64() - uint64(i)
+		latestBlock := db.GetBlockByNumber(latestNumber)
+		db.GetBlockByHash(latestBlock.Hash())
+		db.GetHeaderByHash(latestBlock.Hash())
+		db.GetBlockByHash(latestBlock.Hash())
+		rawdb.ReadBlockRewardAccumulator(db, latestNumber)
+		rawdb.ReadBlockCommitSig(db, latestNumber)
+		for shard := 0; shard < int(params.Sha256BaseGas); shard++ {
```

Comment:
params.Sha256BaseGas is wrong, should be NumShards, right?

Reply:
yes, fixed.